addons/kafka/scripts/kafka-27-server-setup.sh

```diff
@@ -0,0 +1,247 @@
+    return 1
+  fi
+
+  # Todo: currently only nodeport and clusterIp network modes are supported. LoadBalance is not supported yet and needs future support.
+  if [ -n "$advertised_svc_host_value" ] && [ -n "$advertised_svc_port_value" ] && [ "$advertised_svc_port_value" != "9092" ]; then
+    # enable NodePort, use node ip + mapped port as client connection
+    nodeport_domain="${advertised_svc_host_value}:${advertised_svc_port_value}"
+    export KAFKA_CFG_ADVERTISED_LISTENERS="INTERNAL://${current_pod_fqdn}:9094,CLIENT://${nodeport_domain}"
+    echo "[cfg]KAFKA_CFG_ADVERTISED_LISTENERS=$KAFKA_CFG_ADVERTISED_LISTENERS"
+  elif [ "${KB_BROKER_DIRECT_POD_ACCESS}" == "true" ]; then
+    export KAFKA_CFG_ADVERTISED_LISTENERS="INTERNAL://${current_pod_fqdn}:9094,CLIENT://${MY_POD_IP}:9092"
+    echo "[cfg]KAFKA_CFG_ADVERTISED_LISTENERS=$KAFKA_CFG_ADVERTISED_LISTENERS"
+  else
+    # default, use headless service url as client connection
+    export KAFKA_CFG_ADVERTISED_LISTENERS="INTERNAL://${current_pod_fqdn}:9094,CLIENT://${current_pod_fqdn}:9092"
+    echo "[cfg]KAFKA_CFG_ADVERTISED_LISTENERS=$KAFKA_CFG_ADVERTISED_LISTENERS"
+  fi
+
+  # override node.id setting
+  # increments based on a specified base to avoid conflicts with controller settings
+  INDEX=$(echo $MY_POD_NAME | grep -o "\-[0-9]\+\$")
+  INDEX=${INDEX#-}
+  BROKER_NODE_ID=$(( $INDEX + $BROKER_MIN_NODE_ID ))
+  export KAFKA_CFG_NODE_ID="$BROKER_NODE_ID"
+  export KAFKA_CFG_BROKER_ID="$BROKER_NODE_ID"
+  echo "[cfg]KAFKA_CFG_NODE_ID=$KAFKA_CFG_NODE_ID"
+}
+
+set_zookeeper_connect() {
+    # Check if KB_KAFKA_ZOOKEEPER_CONN is set
+    if [[ -z "$KB_KAFKA_ZOOKEEPER_CONN" ]]; then
+        echo "Error: KB_KAFKA_ZOOKEEPER_CONN is not set"
+        return 1
+    fi
+
+    if [ -n "$KB_KAFKA_ZK_SUB_PATH" ]; then
+      # Set KAFKA_CFG_ZOOKEEPER_CONNECT to the concat of KB_KAFKA_ZOOKEEPER_CONN and KB_KAFKA_ZK_SUB_PATH
+      export KAFKA_CFG_ZOOKEEPER_CONNECT="$KB_KAFKA_ZOOKEEPER_CONN/$KB_KAFKA_ZK_SUB_PATH"
+    else
+      # Set KAFKA_CFG_ZOOKEEPER_CONNECT to the value of KB_KAFKA_ZOOKEEPER_CONN
+      export KAFKA_CFG_ZOOKEEPER_CONNECT="$KB_KAFKA_ZOOKEEPER_CONN"
+    fi
+
+    # Optionally, print the value to verify
+    echo "[cfg]export KAFKA_CFG_ZOOKEEPER_CONNECT=$KAFKA_CFG_ZOOKEEPER_CONNECT,for kafka-server."
+}
+
+start_server() {
+  load_common_library
+  set_tls_configuration_if_needed
+  convert_server_properties_to_env_var
+  override_sasl_configuration
+  set_jvm_configuration
+  set_zookeeper_connect
+  set_cfg_metadata
+
+  exec /entrypoint.sh /run.sh
+}
+
+# This is magic for shellspec ut framework.
+# Sometime, functions are defined in a single shell script.
+# You will want to test it. but you do not want to run the script.
+# When included from shellspec, __SOURCED__ variable defined and script
+# end here. The script path is assigned to the __SOURCED__ variable.
+${__SOURCED__:+false} : || return 0
+
+# main
+start_server
```

Comment:
Are there any significant differences in the startup method between Kafka 2.7 and the current versions? Is it necessary to maintain a separate startup script for 2.7? Can we reuse the existing one?

Reply:
Not too significant differences, but i think this new script maybe useful for kafka 2.x, and a separate script can keep the logic clean between 2.x and 3.x